# HG changeset patch
# User Karel Zak <kzak@redhat.com>
# Date 1454171239 0
#      Sat Jan 30 16:27:19 2016 +0000
# Node ID 4f11a04b18c224c600f171148b8a9a573fc954a9
# Parent  405c15955db244cd169dbe0634ce07deff79d531
[PATCH] feature: quasi-delete
Mark emails that should be hidden, but not deleted
---
 OPS         |  1 +
 curs_main.c | 14 ++++++++++++++
 functions.h |  2 ++
 mutt.h      |  1 +
 mx.c        |  5 +++--
 5 files changed, 21 insertions(+), 2 deletions(-)

diff -r 405c15955db2 -r 4f11a04b18c2 OPS
--- a/OPS	Mon Jan 25 17:02:57 2016 +0000
+++ b/OPS	Sat Jan 30 16:27:19 2016 +0000
@@ -128,6 +128,7 @@ OP_MAIN_ROOT_MESSAGE "jump to root messa
 OP_MAIN_SET_FLAG "set a status flag on a message"
 OP_MAIN_SYNC_FOLDER "save changes to mailbox"
 OP_MAIN_TAG_PATTERN "tag messages matching a pattern"
+OP_MAIN_QUASI_DELETE "delete from mutt, don't touch on disk"
 OP_MAIN_UNDELETE_PATTERN "undelete messages matching a pattern"
 OP_MAIN_UNTAG_PATTERN "untag messages matching a pattern"
 OP_MARK_MSG "create a hotkey macro for the current message"
diff -r 405c15955db2 -r 4f11a04b18c2 PATCHES
--- a/PATCHES	Mon Jan 25 17:02:57 2016 +0000
+++ b/PATCHES	Sat Jan 30 16:27:19 2016 +0000
@@ -1,3 +1,4 @@
+patch-quasi-delete-neomutt
 patch-index-color-neomutt
 patch-status-color-neomutt
 patch-skip-quoted-neomutt
diff -r 405c15955db2 -r 4f11a04b18c2 curs_main.c
--- a/curs_main.c	Mon Jan 25 17:02:57 2016 +0000
+++ b/curs_main.c	Sat Jan 30 16:27:19 2016 +0000
@@ -1293,6 +1293,20 @@ int mutt_index_menu (void)
 	  menu->redraw = REDRAW_FULL;
 	break;
 
+      case OP_MAIN_QUASI_DELETE:
+	if (tag) {
+	  for (j = 0; j < Context->vcount; j++) {
+	    if (Context->hdrs[Context->v2r[j]]->tagged) {
+	      Context->hdrs[Context->v2r[j]]->quasi_deleted = TRUE;
+	      Context->changed = TRUE;
+	    }
+	  }
+	} else {
+	  CURHDR->quasi_deleted = TRUE;
+	  Context->changed = 1;
+	}
+	break;
+
 #ifdef USE_SIDEBAR
       case OP_SIDEBAR_OPEN:
 #endif
diff -r 405c15955db2 -r 4f11a04b18c2 functions.h
--- a/functions.h	Mon Jan 25 17:02:57 2016 +0000
+++ b/functions.h	Sat Jan 30 16:27:19 2016 +0000
@@ -170,6 +170,7 @@ const struct binding_t OpMain[] = { /* m
   { "mail-key",			OP_MAIL_KEY,			"\033k" },
   { "decrypt-copy",		OP_DECRYPT_COPY,		NULL },
   { "decrypt-save",		OP_DECRYPT_SAVE,		NULL },
+  { "quasi-delete",		OP_MAIN_QUASI_DELETE,		NULL },
 
 #ifdef USE_SIDEBAR
   { "sidebar-next",		OP_SIDEBAR_NEXT,		NULL },
@@ -288,6 +289,7 @@ const struct binding_t OpPager[] = { /* 
   { "decrypt-save",    	OP_DECRYPT_SAVE,		NULL },
 
   { "what-key",		OP_WHAT_KEY,		NULL },
+  { "quasi-delete",	OP_MAIN_QUASI_DELETE,		NULL },
 
 #ifdef USE_SIDEBAR
   { "sidebar-next",		OP_SIDEBAR_NEXT,		NULL },
diff -r 405c15955db2 -r 4f11a04b18c2 mutt.h
--- a/mutt.h	Mon Jan 25 17:02:57 2016 +0000
+++ b/mutt.h	Sat Jan 30 16:27:19 2016 +0000
@@ -764,6 +764,7 @@ typedef struct header
   unsigned int tagged : 1;
   unsigned int deleted : 1;
   unsigned int purge : 1;               /* skip trash folder when deleting */
+  unsigned int quasi_deleted : 1;	/* deleted from mutt, but not modified on disk */
   unsigned int changed : 1;
   unsigned int attach_del : 1; 		/* has an attachment marked for deletion */
   unsigned int old : 1;
diff -r 405c15955db2 -r 4f11a04b18c2 mx.c
--- a/mx.c	Mon Jan 25 17:02:57 2016 +0000
+++ b/mx.c	Sat Jan 30 16:27:19 2016 +0000
@@ -1033,9 +1033,10 @@ void mx_update_tables(CONTEXT *ctx, int 
 #define this_body ctx->hdrs[j]->content
   for (i = 0, j = 0; i < ctx->msgcount; i++)
   {
-    if ((committing && (!ctx->hdrs[i]->deleted || 
+    if (!ctx->hdrs[i]->quasi_deleted &&
+	((committing && (!ctx->hdrs[i]->deleted ||
 			(ctx->magic == MUTT_MAILDIR && option (OPTMAILDIRTRASH)))) ||
-	(!committing && ctx->hdrs[i]->active))
+	(!committing && ctx->hdrs[i]->active)))
     {
       if (i != j)
       {
